File: backend/BackendApi/Models/Book.cs
namespace BackendApi.Models;

/// <summary>
/// Represents a single book record in the Bookstore database.
/// Each property maps directly to a column in the Books table of Bookstore.sqlite.
/// Entity Framework Core uses these property names to match the database columns.
/// </summary>
public class Book
{
    /// <summary>
    /// Primary key — auto-incremented by the database (AUTOINCREMENT in SQLite).
    /// EF Core recognizes "BookID" as the primary key by convention.
    /// </summary>
    public int BookID { get; set; }

    /// <summary>The full title of the book (e.g. "Les Miserables").</summary>
    public string Title { get; set; } = string.Empty;

    /// <summary>The author's full name (e.g. "Victor Hugo").</summary>
    public string Author { get; set; } = string.Empty;

    /// <summary>The publishing company (e.g. "Signet").</summary>
    public string Publisher { get; set; } = string.Empty;

    /// <summary>
    /// The International Standard Book Number, stored as a string so leading
    /// zeros and hyphens are preserved (e.g. "978-0451419439").
    /// </summary>
    public string ISBN { get; set; } = string.Empty;

    /// <summary>
    /// Broad classification — typically "Fiction" or "Non-Fiction".
    /// Stored in the Classification column of the database.
    /// </summary>
    public string Classification { get; set; } = string.Empty;

    /// <summary>
    /// More specific genre/category (e.g. "Biography", "Classic", "Historical").
    /// Stored in the Category column of the database.
    /// </summary>
    public string Category { get; set; } = string.Empty;

    /// <summary>Total number of pages in the book.</summary>
    public int PageCount { get; set; }

    /// <summary>Retail price in US dollars (e.g. 9.95).</summary>
    public double Price { get; set; }
}

File: frontend/src/types/Book.ts
/**
 * Book and API response types — frontend mirror of BackendApi.Models.Book
 * and the JSON envelope from GET /api/books.
 *
 * Note: ASP.NET Core serializes property names in camelCase by default,
 * so "BookID" becomes "bookID", "PageCount" becomes "pageCount", etc.
 * These names must match exactly what the API sends.
 */
export interface Book {
  bookID: number;         // Primary key (auto-incremented in the database)
  title: string;          // Full book title
  author: string;         // Author's name
  publisher: string;      // Publishing company
  isbn: string;           // International Standard Book Number
  classification: string; // Broad type: "Fiction" or "Non-Fiction"
  category: string;       // Specific genre: "Biography", "Classic", etc.
  pageCount: number;      // Total number of pages
  price: number;          // Retail price in USD
}

/**
 * Shape of the JSON object returned by GET /api/books.
 * Includes the page of books AND the pagination metadata needed to render
 * the Previous / Next / page-number controls.
 * Also includes the active category so the frontend can reflect it.
 */
export interface BooksResponse {
  books: Book[];       // The books on the current page
  totalBooks: number;  // Total books matching the current filter
  page: number;        // The current 1-based page number
  pageSize: number;    // How many books are shown per page
  totalPages: number;  // Total number of pages (Math.ceil(totalBooks / pageSize))
  category: string;    // The active category filter (empty = all)
}
